File: 104-maximum-depth-of-binary-tree/solution.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

type Traverse = { node: TreeNode; depth: number };

function maxDepth(root: TreeNode | null): number {
  if (!root) return 0;
  const traversal: Traverse[] = [{ node: root, depth: 1 }];
  let answer = 0;
  while (traversal[0]) {
    const { node, depth } = traversal.shift() as Traverse;
    answer = answer > depth ? answer : depth;
    if (node.left) traversal.push({ node: node.left, depth: depth + 1 });
    if (node.right) traversal.push({ node: node.right, depth: depth + 1 });
  }
  return answer;
}

File: 104-maximum-depth-of-binary-tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void traversal(int& ans, int curr, TreeNode* node) {
        if(node == NULL) return;
        ans = max(ans, curr);
        traversal(ans, curr + 1, node->left);
        traversal(ans, curr + 1, node->right);
    }

    int maxDepth(TreeNode* root) {
        int ans = 0;
        traversal(ans, 1, root);
        return ans;
    }
};
